jsapp/js/components/projectDownloads/anonymousExports.es6

```diff
@@ -0,0 +1,164 @@
+      isPending: false,
+      exportUrl: null,
+    };
+    this.unlisteners = [];
+    this.fetchIntervalId = null;
+    autoBind(this);
+  }
+
+  componentDidMount() {
+    this.unlisteners.push(
+      exportsStore.listen(this.onExportsStoreChange),
+      actions.exports.createExport.completed.listen(this.onCreateExportCompleted),
+      actions.exports.getExport.completed.listen(this.onGetExportCompleted),
+    );
+  }
+
+  componentWillUnmount() {
+    this.unlisteners.forEach((clb) => {clb();});
+  }
+
+  onExportsStoreChange() {
+    this.setState({
+      selectedExportType: exportsStore.getExportType(),
+      exportUrl: null,
+    });
+  }
+
+  onCreateExportCompleted(exportData) {
+    this.fetchExport(exportData.uid);
+  }
+
+  onGetExportCompleted(exportData) {
+    this.prepareFetchInterval(exportData.uid, exportData.status);
+
+    if (exportData.status === EXPORT_STATUSES.complete) {
+      this.setState({
+        isPending: false,
+        exportUrl: exportData.result,
+      });
+      this.downloadUrl(this.state.exportUrl);
+    }
+  }
+
+  onSubmit() {
+    if (this.state.exportUrl) {
+      // we remember the current type download to not make multiple calls
+      this.downloadUrl(this.state.exportUrl);
+    } else {
+      this.setState({isPending: true});
+
+      const defaultExportFormat = getContextualDefaultExportFormat(this.props.asset);
+
+      // NOTE: this wouldn't work for legacy formats, but luckily we don't allow
+      // choosing legacy types in this component
+      actions.exports.createExport(
+        this.props.asset.uid,
+        {
+          type: this.state.selectedExportType.value,
+          fields_from_all_versions: DEFAULT_EXPORT_SETTINGS.INCLUDE_ALL_VERSIONS,
+          group_sep: DEFAULT_EXPORT_SETTINGS.GROUP_SEPARATOR,
+          hierarchy_in_labels: DEFAULT_EXPORT_SETTINGS.INCLUDE_GROUPS,
+          lang: defaultExportFormat.value,
+          multiple_select: DEFAULT_EXPORT_SETTINGS.EXPORT_MULTIPLE.value,
+        }
+      );
+    }
+  }
+
+  prepareFetchInterval(exportUid, exportStatus) {
+    // if the export is not complete yet, and there is no fetch interval
+    // fetch it in some time again and again
+    if (
+      exportStatus !== EXPORT_STATUSES.error &&
+      exportStatus !== EXPORT_STATUSES.complete &&
+      !this.fetchIntervalId
+    ) {
+      const intervalId = setInterval(this.fetchExport.bind(this, exportUid), EXPORT_REFRESH_TIME);
+      this.fetchIntervalId = intervalId;
+    }
+
+    // clean up after it is completed
+    if (
+      exportStatus === EXPORT_STATUSES.error ||
+      exportStatus === EXPORT_STATUSES.complete
+    ) {
+      this.removeFetchInterval();
+    }
+  }
+
+  removeFetchInterval() {
+    clearInterval(this.fetchIntervalId);
+    this.fetchIntervalId = null;
+  }
+
+  fetchExport(exportUid) {
+    actions.exports.getExport(this.props.asset.uid, exportUid);
+  }
+
+  downloadUrl(url) {
+    const aEl = document.createElement('a');
```

Comment:
We disable the submit button when user clicks, so it will not trigger multiple times at once. As for multiple consecutive times (click submit, wait for file to be ready, click again, repeat) - this function is creating the `aEl` in void, it doesn't attach it to any existing document node, so garbage collection will clean it up :)

Reply:
Oh interesting, I didn't know creating elements like like get picked up by the garbage collector--I just assumed it was appended invisibly to the DOM :sweat_smile: